File: android/src/main/kotlin/fl/channel/FlDataStream.kt
package fl.channel

typealias FlDataStreamHandler<T> = (T) -> Unit

// 自定义数据流管理器
class FlDataStream<T> {

    private val dataHandlers: MutableList<FlDataStreamHandler<T>> = ArrayList()

    // 发送数据
    fun send(data: T) {
        for (handler in dataHandlers) {
            handler(data)
        }
    }

    // 监听数据
    fun listen(handler: FlDataStreamHandler<T>): () -> Unit {
        dataHandlers.add(handler)
        return { cancel(handler) }
    }

    // 取消监听
    private fun cancel(handler: FlDataStreamHandler<T>) {
        dataHandlers.remove(handler)
    }

}

File: ios/Classes/FlDataStream.swift
import Foundation

typealias FlDataStreamHandler<T> = (T) -> Void

// 定义回调方法类型的包装类
class FlDataStreamHandlerWrapper<T> {
    let handler: FlDataStreamHandler<T>

    init(handler: @escaping (T) -> Void) {
        self.handler = handler
    }
}

// 自定义数据流管理器
class FlDataStream<T> {
    private var dataHandlers: [FlDataStreamHandlerWrapper<T>] = []

    // 发送数据
    func send(data: T) {
        for wrapper in dataHandlers {
            wrapper.handler(data)
        }
    }

    // 监听数据
    @discardableResult
    func listen(handler: @escaping FlDataStreamHandler<T>) -> () -> Void {
        let wrapper = FlDataStreamHandlerWrapper<T>(handler: handler)
        dataHandlers.append(wrapper)
        return { [weak self] in
            self?.cancel(wrapper: wrapper)
        }
    }

    // 取消监听
    private func cancel(wrapper: FlDataStreamHandlerWrapper<T>) {
        if let index = dataHandlers.firstIndex(where: { $0 === wrapper }) {
            dataHandlers.remove(at: index)
        }
    }
}